sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/sycl_ext_oneapi_bindless_images.asciidoc

```diff
@@ -0,0 +1,541 @@
+`destroy_image_handle`.
+
+=== Explicit copies
+
+```cpp
+namespace sycl {
+class queue {
+public:
+  event ext_image_memcpy(ext::oneapi::image_mem_handle Dest, void *Src,
+                         const ext::oneapi::image_descriptor &Desc);
+  event ext_image_memcpy(ext::oneapi::image_mem_handle Dest, void *Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         event DepEvent);
+  event ext_image_memcpy(ext::oneapi::image_mem_handle Dest, void *Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         const std::vector<event> &DepEvents);
+
+  event ext_image_memcpy(void *Dest, ext::oneapi::image_mem_handle Src,
+                         const ext::oneapi::image_descriptor &Desc);
+  event ext_image_memcpy(void *Dest, ext::oneapi::image_mem_handle Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         event DepEvent);
+  event ext_image_memcpy(void *Dest, ext::oneapi::image_mem_handle Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         const std::vector<event> &DepEvents);
+
+  event ext_image_memcpy(void *Dest, void *Src, 
+                         const ext::oneapi::image_descriptor &Desc);
+  event ext_image_memcpy(void *Dest, void *Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         event DepEvent);
+  event ext_image_memcpy(void *Dest, void *Src,
+                         const ext::oneapi::image_descriptor &Desc,
+                         const std::vector<event> &DepEvents);
+};
+}
+```
+
+An `ext_image_memcpy` function is proposed as a method of the queue, to 
+copy image memory. It can be used to copy image memory from host to device, or 
+device to host. The image descriptor is used to determine the dimensions of 
+the image, and the image row pitch, when copying the data.
+
+For images allocated using USM, existing SYCL functionality can used to 
+copy their memory, but we also provide `ext_image_copy` functions that take 
+USM pointers.
+
+=== Reading and writing inside the kernel
+
+```cpp
+namespace sycl::ext::oneapi {
+
+DataT read_image(const unsampled_image_handle &imageHandle,
+                 const CoordT &coords);
+DataT read_image(const sampled_image_handle &imageHandle, 
+                 const CoordT &coords);
+
+template <typename DataT, typename CoordT>
+void write_image(const unsampled_image_handle &imageHandle,
+                 const CoordT &Coords, const DataT &Color);
+
+}
+```
+
+Inside a kernel it's possible to read an image using `read_image` via the image 
+handle, which returns a value of the user specified type `DataT`.
+Similarly, it's possible to write to an image via a handle using `write_image`.
+
+`write_image` is not available for sampled images, or images created from USM 
+memory.
+
+`DataT` must correspond to the type specified in the `image_descriptor`
```

Comment:
There is no `DataT` in the image descriptor.  There are the `image_channel_order` and `image_channel_type` enumerations, but how will they map to a DataT ? 
The reason I ask is that in SYCL 1.2.1 images, no matter the choices for image_channel_order and image_channel_type, the accessor data type was always just `cl_int4` or `cl_float4`  . IIRC, even one channel images use a `cl_int4`, and similarly  there is no difference in the accessor data type for `image_channel_type::unsigned_int8` versus `::unsigned_int16` or `::unsigned_int32`.   They all used `cl_int4` for the accessor type.

Reply:
I wasn't aware this was the case in SYCL 1.2.1. With this proposal we can pass a `DataT` that corresponds to any of the combinations of `image_channel_order` and `image_channel_type`, e.g. an `image_channel_type::fp16` combined with a `iamge_channel_order::rg` would make `DataT == sycl::half2 == sycl::vec<sycl::half, 2>`.

We will need to clarify that we are looking for `sycl::vec` types for multi-channel stores as it is not explained, and maybe include a table of the acceptable combinations and resulting types. We also need to add a restrictions on channels, as CUDA only supports 1, 2, or 4 channel images, when we pass a channel order like `rgb`, we would need to emulate it with a 4-channel image, or get rid of the 3-channel image orders altogether.